website/src/hooks/tasks/useGenericTaskAPI.tsx

```diff
@@ -1,26 +1,43 @@
 import { useState } from "react";
 import { get, post } from "src/lib/api";
-import { BaseTask, TaskResponse, TaskType as TaskTypeEnum } from "src/types/Task";
+import { TaskApiHook } from "src/types/Hooks";
+import { BaseTask, TaskAvailableResponse, TaskResponse, TaskType as TaskTypeEnum } from "src/types/Task";
 import useSWRImmutable from "swr/immutable";
 import useSWRMutation from "swr/mutation";
 
-export const useGenericTaskAPI = <TaskType extends BaseTask>(taskType: TaskTypeEnum) => {
-  type ConcreteTaskResponse = TaskResponse<TaskType>;
+export const useGenericTaskAPI = <TaskType extends BaseTask>(taskType: TaskTypeEnum): TaskApiHook<TaskType> => {
+  const [response, setReponse] = useState<TaskResponse<TaskType>>({ status: "AWAITING_INITIAL" });
+  const [isLoading, setIsLoading] = useState(true);
 
-  const [tasks, setTasks] = useState<ConcreteTaskResponse[]>([]);
+  const { mutate: requestNewTask } = useSWRImmutable<TaskAvailableResponse<TaskType>>(
+    "/api/new_task/" + taskType,
+    get,
+    {
+      onSuccess: (response) => {
+        setIsLoading(false);
+        setReponse({ status: "AVAILABLE", ...response });
+      },
+      onError: () => {
+        // We could check for code 503 here for truely unavailable, but we need to do something with other errors anyway.
+        setIsLoading(false);
+        setReponse({ status: "NONE_AVAILABLE" });
+      },
+      revalidateOnMount: true,
+      dedupingInterval: 500,
+    }
+  );
 
-  const { isLoading, mutate, error } = useSWRImmutable<ConcreteTaskResponse>("/api/new_task/" + taskType, get, {
-    onSuccess: (data) => setTasks([data]),
-    revalidateOnMount: true,
-    dedupingInterval: 500,
-  });
-
-  const { trigger } = useSWRMutation("/api/update_task", post, {
-    onSuccess: async (newTask: ConcreteTaskResponse) => {
-      setTasks((oldTasks) => [...oldTasks, newTask]);
-      mutate();
+  const { trigger: completeTask } = useSWRMutation<TaskAvailableResponse<TaskType>>("/api/update_task", post, {
+    onSuccess: () => {
+      setIsLoading(true);
+      requestNewTask();
```

Comment:
no set status here?

Reply:
Correct, the data returned from /api/update_task is just a "task" with the task type "done" (or some equivalent, didn't check the wording), we don't actually want to try to show such a non-task task and so we just need to request a new task.